cf-custom-resources/lib/nlb-cert-validator-updater.js

```diff
@@ -0,0 +1,301 @@
+                await activate(options, certificateARN, loadBalancerDNS, loadBalancerHostedZoneID);
+                break;
+            case "Update":
+            case "Delete":
+            default:
+                throw new Error(`Unsupported request type ${event.RequestType}`);
+        }
+    };
+
+    try {
+        await handler();
+        await report(event, context, "SUCCESS", physicalResourceID);
+    } catch (err) {
+        console.log(`Caught error for service ${serviceName}: ${err.message}`);
+        await report(event, context, "FAILED", physicalResourceID, null, err.message);
+    }
+};
+
+/**
+ * Validate that the aliases are not in use.
+ *
+ * @param {Set<String>} aliases for the service.
+ * @param {String} loadBalancerDNS the DNS of the service's load balancer.
+ * @throws error if at least one of the aliases is not valid.
+ */
+async function validateAliases(aliases, loadBalancerDNS) {
+    for (let alias of aliases) {
+        const data = await envRoute53.listResourceRecordSets({
+            HostedZoneId: envHostedZoneID,
+            MaxItems: "1",
+            StartRecordName: alias,
+        }).promise();
+
+        let recordSet = data["ResourceRecordSets"];
+        if (!recordSet || recordSet.length === 0) {
+            continue; // The alias record does not exist in the hosted zone, hence valid.
+        }
+        if (recordSet[0].AliasTarget && recordSet[0].AliasTarget.DNSName === `${loadBalancerDNS}.`) {
+            continue; // The record is an alias record and is in use by myself, hence valid.
+        }
+        throw new Error(`alias ${alias} is in use`);
+    }
+}
+
+/**
+ * Requests a public certificate from AWS Certificate Manager, using DNS validation.
+ *
+ * @param {Set<String>} aliases the subject alternative names for the certificate.
+ * @param {String} physicalResourceID
+ * @return {String} The ARN of the requested certificate.
+ */
+async function requestCertificate(aliases, physicalResourceID) {
+    const { CertificateArn } =await acm.requestCertificate({
+        DomainName: certificateDomain,
+        IdempotencyToken: physicalResourceID,
+        SubjectAlternativeNames: aliases.size === 0? null: [...aliases],
+        Tags: [
+            {
+                Key: "copilot-application",
+                Value: appName,
+            },
+            {
+                Key: "copilot-environment",
+                Value: envName,
+            },
+        ],
+        ValidationMethod: "DNS",
+    }).promise();
+    return CertificateArn;
+}
+
+/**
+ * Wait until the validation options are ready
+ *
+ * @param certificateARN
+ * @param {Set<String>} aliases for the service.
+ */
+async function waitForValidationOptionsToBeReady(certificateARN, aliases) {
+    let expectedCount = aliases.size + 1; // Expect one validation option for each alias and the cert domain.
+
+    let attempt; // TODO: This wait loops could be further abstracted.
```

Comment:
Yes in a following PR - especially because DELETE also needs to have a wait loop like this. I think it may improve code's readability if this can be abstracted.